### PR TITLE
Preparations for sending emails.

### PR DESCRIPTION
**Pull request by isoos**
- Extracted and renamed `Author` to `EmailAddress`.
- Created `EmailMessage` and `createPackageUploadedEmail`
- Long e-mail body lines will be split by 72-character long lines.
- #1749

### CODE REVIEW

File: app/lib/shared/email.dart [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    bodyText.writeln('The listed authors are the following:');
+    authors.forEach((author) {
+      bodyText.writeln('- $author');
+    });
+    bodyText.writeln();
+  }
+
+  bodyText
+    ..writeln(
+        'If you think this is a mistake or fraud, contact us at ${_defaultFrom.email}')
```
Comment (jonasfj) [inline app/lib/shared/email.dart]: don't write that email, file an issue on github:
https://github.com/dart-lang/pub-dartlang-dart/issues

File: app/lib/shared/email.dart [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  final String email;
+
+  EmailAddress(this.name, this.email);
+
+  factory EmailAddress.parse(String value) {
+    value = value.trim();
+    String name = value;
+    String email;
+
+    final match = _nameEmailRegExp.matchAsPrefix(value);
```
Comment (jonasfj) [inline app/lib/shared/email.dart]: Couldn't we propose this parse string to email, and reflow body text as static method in the `mailer` package?

Or at-least use the types defined in the mailer package, that way we don't have to wrap back and forth..?

File: app/lib/shared/email.dart [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+/// Creates the [EmailMessage]s that we be sent on new package upload.
+EmailMessage createPackageUploadedEmail({
+  @required String packageName,
+  @required String packageVersion,
+  @required String uploaderEmail,
+  @required List<EmailAddress> authorizedUploaders,
+  @required List<EmailAddress> authors,
+}) {
+  final subject = 'Package upload on pub: $packageName $packageVersion';
+  final bodyText = new StringBuffer()
```
Comment (jonasfj) [inline app/lib/shared/email.dart]: Couldn't we just use a multi-line string in dart?

File: app/lib/shared/email.dart [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    ..writeln('Dear package maintainer,')
+    ..writeln()
+    ..writeln(
+        '$uploaderEmail uploaded a new version of package $packageName: $packageVersion')
+    ..write('\n');
```
Comment (jonasfj) [inline app/lib/shared/email.dart]: `writeln()`

File: app/lib/shared/email.dart [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+  void addRecipient(EmailAddress address) {
+    if (address.email == null || addedEmails.contains(address.email)) {
+      return;
+    }
+    addedEmails.add(address.email);
+    recipients.add(address);
+  }
+
+  authorizedUploaders.forEach(addRecipient);
```
Comment (jonasfj) [inline app/lib/shared/email.dart]: Or we could do this with:
```dart
final emails = Set<String>();  // avoid duplicate recipients
final recipients = authorizedUploaders.where((u) => u.email != null && emails.add(u)).toList();
```

Or perhaps we could move this logic to `mailer` with a PR.

File: app/lib/shared/email.dart (hunk 2 above)
Comment (isoos) [reply app/lib/shared/email.dart]: `EmailAddress` (previously: `Author`) is something we use in our code outside of the scope of email. I wouldn't rely on a third-party package to provide the data structure for it. Besides: I haven't seen mailer to parse the string, it only has a structure to hold it.

`EmailMessage` on the other hand could be replaced.

File: app/lib/shared/email.dart (hunk 3 above)
Comment (isoos) [reply app/lib/shared/email.dart]: Yeah, it started out as a bit more complex template, now a multi-line string could be just fine. Changing it...

File: app/lib/shared/email.dart (hunk 5 above)
Comment (isoos) [reply app/lib/shared/email.dart]: I've filed and issue in the `mailer` package and added a TODO here to later clear this up. With the current logic, we could just use the `authorizedUploaders` list without any further processing. Do you want to keep this filtering here, or shall I simplify that part too?

Comment (jonasfj) [reply app/lib/shared/email.dart]: I'm okay either way :)

Comment (isoos) [reply app/lib/shared/email.dart]: Simplicity won :)

**Pull request merged**